lib/src/main/java/com/basistheory/android/view/mask/MaskWatcher.kt

```diff
@@ -0,0 +1,48 @@
+package com.basistheory.android.view.mask
+
+import android.text.Editable
+import android.text.TextWatcher
+
+internal class MaskWatcher(mask: List<Any>) : TextWatcher {
+    private val mask: Mask = Mask(mask)
+
+    private var selfChange: Boolean = false
+    private var result: MaskResult? = null
+
+    val maskedValue: String
+        get() = result?.maskedValue.orEmpty()
+
+    val unMaskedValue: String
+        get() = result?.unMaskedValue.orEmpty()
+
+    val isDone: Boolean
+        get() = result?.isDone ?: false
+
+    val isDirty: Boolean
+        get() = result?.isDirty ?: false
+
+    override fun afterTextChanged(editable: Editable?) {
+        if (selfChange || editable.isNullOrEmpty()) return
+
+        selfChange = true
+        result?.apply(editable)
+        selfChange = false
+    }
```

Comment:
why split logic between two event handlers? Can we not do all of this in the same event hander like `onTextChanged`?

Reply:
`result.apply` temporarily clears existing [text filters](https://developer.android.com/reference/android/text/InputFilter)  in order to replace the input value with the masked value and those filters can only be accessed via `Editable` not `CharSequence`.